source/mesh_deformation/fastscape.cc

```diff
@@ -0,0 +1,1739 @@
+      if (Utilities::MPI::this_mpi_process(this->get_mpi_communicator()) == 0)
+        {
+          // Initialize the variables that will be sent to FastScape.
+          std::vector<double> h(array_size, std::numeric_limits<double>::max());
+          std::vector<double> vx(array_size);
+          std::vector<double> vy(array_size);
+          std::vector<double> vz(array_size);
+          std::vector<double> kf(array_size);
+          std::vector<double> kd(array_size);
+          std::vector<double> b(array_size);
+          std::vector<double> sf(array_size);
+          std::vector<double> h_old(array_size);
+
+          fill_fastscape_arrays(h,
+                                kd,
+                                kf,
+                                vx,
+                                vy,
+                                vz,
+                                temporary_variables);
+
+          if (current_timestep == 1 || restart)
+            {
+              this->get_pcout() << "   Initializing FastScape... " << (1+maximum_surface_refinement_level+additional_refinement) <<
+                                " levels, cell size: " << dx << " m." << std::endl;
+
+              // Set ghost nodes before initializing.
+              if (use_ghost_nodes && !restart)
+                set_ghost_nodes(h,
+                                vx,
+                                vy,
+                                vz,
+                                true);
+
+              // If we are restarting from a checkpoint, load h values for FastScape instead of using the ASPECT values.
+              if (restart)
+                {
+                  read_restart_files(h,
+                                     b,
+                                     sf);
+
+                  restart = false;
+                }
+
+              initialize_fastscape(h,
+                                   b,
+                                   kd,
+                                   kf,
+                                   sf);
+            }
+          else
+            {
+              // If it isn't the first timestep we ignore initialization and instead copy all height values from FastScape.
+              // Generally, we overwrite the topography data from ASPECT as FastScape may be at a higher resolution. However,
+              // if we are not using FastScape to advect then we do not want to do this and instead use the ASPECT values.
+              if (use_velocities)
+                fastscape_copy_h_(h.data());
+            }
+
+          // Find the appropriate sediment rain based off the time interval.
+          const double time_in_years = this->get_time() / year_in_seconds;
+          double sediment_rain = sediment_rain_rates[0];
+          for (unsigned int j=0; j<sediment_rain_times.size(); ++j)
+            {
+              if (time_in_years > sediment_rain_times[j])
+                sediment_rain = sediment_rain_rates[j+1];
+            }
```

Comment:
I've added an assert to make sure it is increasing. What would be a good way to shortcut this?

Reply:
The assertion should go into the function that reads the parameters, so that the error is shown as early as possible if a user uses non-ascending times.

To make this code shorter, you might want to use `std::lower_bound()`, which uses a binary search through the array.